include/networkit/graph/Graph.hpp

```diff
@@ -139,6 +145,389 @@ class Graph final {
     //!< same schema (and same order!) as outEdges
     std::vector<std::vector<edgeid>> outEdgeIds;
 
+private:
+    // user attributes of this graph (node attributes so far)
+    // Base class for all node attributes.
+    class NodeAttributeStorageBase {
+    public:
+        NodeAttributeStorageBase(const Graph *graph, std::string name, std::type_index type)
+            : name{std::move(name)}, type{type}, theGraph{graph} {}
+
+        virtual ~NodeAttributeStorageBase() = default;
+
+        virtual void invalidateAttributes() = 0;
+
+        std::string getName() { return name; }
+
+        std::type_index getType() { return type; }
+
+        bool isValid(index i) { return i < valid.size() && valid[i]; }
+
+        // Called by Graph when node is deleted.
+        void invalidate(index i) {
+            if (isValid(i)) {
+                valid[i] = false;
+                --validElements;
+            }
+        }
+
+    protected:
+        void markValid(index i) {
+            if (!theGraph->hasNode(i)) {
+                throw std::runtime_error("This node does not exist");
+            }
```

Comment:
I think that we should expect the user to pass correct indices and replace these bound checks with assertions.

Reply:
The outcome of the discussion here: https://github.com/networkit/networkit/discussions/769 is that `Graph` should use exceptions, not assertion to check for broken indices.